Add signature checking to install-debs.py

akoeplinger opened the pull request:
cc @am11

akoeplinger commented on eng/common/cross/install-debs.py:
```diff
@@ -63,10 +71,9 @@ async def download_package_index_parallel(mirror, arch, suites):
+                tasks.append(fetch_and_decompress(session, mirror, arch, suite, component))
 
-        results = await asyncio.gather(*tasks, return_exceptions=True)
+        results = await asyncio.gather(*tasks)
```
I noticed that it was suppressing exceptions if this was set

akoeplinger commented on eng/common/cross/install-debs.py:
```diff
@@ -225,11 +286,11 @@ def install_packages(mirror, packages_info, aliases, tmp_dir, extract_dir, ar_to
         deb_file = package_to_deb_file_map.get(pkg)
         if deb_file and os.path.exists(deb_file):
-            extract_deb_file_using_dpkg(deb_file, tmp_dir, extract_dir, ar_tool)
+            extract_deb_file(deb_file, tmp_dir, extract_dir, ar_tool)
```
this wasn't using dpkg so I renamed it

am11 commented on eng/common/cross/build-rootfs.sh:
```diff
@@ -797,15 +797,19 @@ elif [[ -n "$__CodeName" ]]; then
         # shellcheck disable=SC2086
         suites="$__CodeName $__DebianSuites $(echo $__UbuntuSuites | xargs -n 1 | xargs -I {} echo -n "$__CodeName-{} ")"
 
+        if [[ "$__SkipSigCheck" == "0" ]]; then
```
We can copy the block on line 820 before its parent block and delete this?

akoeplinger replied:
I thought about it but we'd need to keep the check for __UpdateOptions anyway so it didn't seem worth it

am11 replied:
We can keep __UpdateOptions there as well. Lots are variables are unused in this block. Just makes it bit easy to read the code. Up to you.

am11 replied:
Please double check if we need to clear `__Keyring` like we have line 823.

am11 commented:
Could you apply fully_trusted change here as well? No need for another PR since Azure Linux has 3.12 atm and won't be updating anytime soon given their pace of updates. 😅

am11 commented:
Yea no rush. Images are building fine (with python 3.12) for now.

```sh
$ docker run --rm mcr.microsoft.com/dotnet-buildtools/prereqs:azurelinux-3.0-net10.0-cross-loongarch64 python3 --version

Python 3.12.3
```

am11 commented on eng/common/cross/install-debs.py:
```diff
@@ -77,21 +84,74 @@ async def download_package_index_parallel(mirror, arch, suites):
+        await download_file(session, release_url, release_file.name)
+        await download_file(session, release_gpg_url, release_gpg_file.name)
+
+        if args.keyring != '':
```
Consistency nit: the current convention is to pass args.xyz to the method calls from main. If you prefer global access, we can bring all of them to the same plan.

akoeplinger replied:
Done

am11 commented on eng/common/cross/install-debs.py:
```diff
@@ -77,21 +84,74 @@ async def download_package_index_parallel(mirror, arch, suites):
+        await download_file(session, release_gpg_url, release_gpg_file.name)
+
+        if args.keyring != '':
+            keyring_arg = f"--keyring {args.keyring}"
```
When `__Keyring` is set in bash script, it has this prefix.
https://github.com/dotnet/arcade/blob/2f07a67129c5999efa73c13d1c4827815406a393/eng/common/cross/build-rootfs.sh#L183C17-L183C26

akoeplinger replied:
that's intentional, we pass `--keyring <file>` to install-debs.py which in turn parses that arg and sets the file

am11 commented:
@akoeplinger I think https://github.com/dotnet/arcade/commit/43f2fa5bc36bba8247772650d0affaff991ecb3b is not optional if we want to keep arguments order independent. I'm a bit busy with another build but please double check before merging. `build-rootfs sid loongarch64` vs. `build-rootfs loongarch64 sid` or just "trust me bro ™️" and apply the patch. 😅

akoeplinger commented:
@am11 I looked at the script but I don't see where it'd make a difference. That said I think we can remove __Keyring and just check whether __KeyringFile is set

am11 commented:
@akoeplinger have you tested it with --skipsigcheck with different orders of argument? It matters when we are running it locally vs. in prereqs.

akoeplinger commented:
hm I think I see what you mean. I don't think we can rely on __UbuntuArch when parsing the `sid)` argument or setting __UbuntuRepo in `loongarch64)` since one or the other isn't defined depending on the order

am11 commented:
It will be so that `build-rootfs sid <arch> --skipsigcheck` and `build-rootfs <arch> --skipsigcheck` will work in same way with or without `--skipemulation`. We don't need larger refactoring for it, the patch should cover the case.

am11 commented on eng/common/cross/install-debs.py:
```diff
@@ -51,22 +59,21 @@ async def download_deb_files_parallel(mirror, packages, tmp_dir):
+                tasks.append(fetch_and_decompress(session, mirror, arch, suite, component, check_sig, keyring))
 
-        results = await asyncio.gather(*tasks, return_exceptions=True)
+        results = await asyncio.gather(*tasks)
```
This is intentional. We need to continue on error here otherwise it fails the build. Some index files are optional and may not be available for certain distro.

akoeplinger replied:
that's already handled by fetch_and_decompress when response.status is not 200

am11 replied:
It fails from this script when it doesn't find some index. That's why I added success and error logging here. apt behavior is the same, it warns and continue.

Lets see if it fails https://github.com/dotnet/dotnet-buildtools-prereqs-docker/pull/1310. Local build was failing when I applied your patch. I suggest you open a similar PR to test variations. You can update loongarch with --skipsigcheck (we don't have keys for Debian sid) and riscv without skipping the check (we have keys for Ubuntu noble) to showcase that it is working.

am11 replied:
There are other issues:
https://dev.azure.com/dnceng-public/public/_build/results?buildId=906926&view=logs&s=6884a131-87da-5381-61f3-d7acc3b91d76&j=fc59f0f2-c1bd-58ae-b870-833d1e8a924c

```
 > [builder 3/4] RUN rootfsEnv="/usr/local/share/rootfs" &&     python3 -m venv "$rootfsEnv" &&     "$rootfsEnv/bin/python" -m pip install aiohttp zstandard &&     PYTHON_EXECUTABLE="$rootfsEnv/bin/python" /scripts/eng/common/cross/build-rootfs.sh loongarch64 sid --skipunmount --skipemulation:
12.19 Downloaded http://ftp.ports.debian.org/debian-ports//dists/unreleased/Release at /tmp/tmp9kl7odro
12.19 Downloaded http://ftp.ports.debian.org/debian-ports//dists/unreleased/Release.gpg at /tmp/tmpd_jy2s9z
12.19 Verifying signature of Release with Release.gpg.
12.19 Error fetching http://ftp.ports.debian.org/debian-ports//dists/unreleased/main/binary-loong64/Packages.gz: cannot access local variable 'keyring_arg' where it is not associated with a value
12.19 Downloaded index: http://ftp.ports.debian.org/debian-ports//dists/sid/main/binary-loong64/Packages.gz
12.19 Downloaded http://ftp.ports.debian.org/debian-ports//dists/sid/Release at /tmp/tmphulowxes
12.19 Downloaded http://ftp.ports.debian.org/debian-ports//dists/sid/Release.gpg at /tmp/tmp5h7tvol5
12.19 Verifying signature of Release with Release.gpg.
12.19 Error fetching http://ftp.ports.debian.org/debian-ports//dists/sid/main/binary-loong64/Packages.gz: cannot access local variable 'keyring_arg' where it is not associated with a value
12.19 Error: Package 'build-essential' was not found in the available packages.
------
ERROR: failed to solve: executor failed running [/bin/sh -c rootfsEnv="/usr/local/share/rootfs" &&     python3 -m venv "$rootfsEnv" &&     "$rootfsEnv/bin/python" -m pip install aiohttp zstandard &&     PYTHON_EXECUTABLE="$rootfsEnv/bin/python" /scripts/eng/common/cross/build-rootfs.sh loongarch64 sid --skipunmount --skipemulation]: exit code: 1
Retry 4/5, retrying in 125 seconds...
#0 building with "default" instance using docker driver
```

I will now revert this line from the PR, and highly recommend you open similar PR downstream with that line in LA64 and RA64 net10.0 docksfiles so we have some coverage.

```dockerfile
RUN rm -rf /scripts && git clone https://github.com/dotnet/arcade --single-branch --depth 1 -b akoeplinger-patch-1 /scripts
```

@akoeplinger, this PR is great, my intention with the feedback is we get the local and cloud scenarios right. For local testing, I'm using podman desktop on macOS for loongarch to test without `--skipemulation` and docker&podman deskotp with `--skipemulation` as docker desktop is missing the registration https://github.com/docker/roadmap/discussions/764. I am sharing the use-cases which I was testing locally. Hope my feedback doesn't come across different than the intended spirit. 👍

akoeplinger replied:
@am11 I tried locally and it fails because we don't have `/usr/share/keyrings/debian-archive-keyring.gpg` nor `/usr/share/keyrings/debian-ports-archive-keyring.gpg` in the builder image, I suppose we need to add it to https://github.com/dotnet/dotnet-buildtools-prereqs-docker/blob/c77a9f6b01eff67ba53a9363f949dee7ccde38c5/src/azurelinux/3.0/net10.0/crossdeps-builder/amd64/Dockerfile#L46-L51

Was this fetched by debootstrap before?

> Hope my feedback doesn't come across different than the intended spirit. 👍

No worries, I'm still on vacation till tomorrow so I can't spend a lot of time on this yet 😄

am11 replied:
`--skipsigcheck` should not require the keyring. That's the purpose of this option. If you incorporate the feedback:

* https://github.com/dotnet/arcade/pull/15374#issuecomment-2569952406 (apply that patch)
* undo `        results = await asyncio.gather(*tasks, return_exceptions=True)`

it will resolve both issues.

am11 replied:
`--skipsigcheck` should not require the keyring. That's the purpose of this arg. If you incorporate the feedback:

* https://github.com/dotnet/arcade/pull/15374#issuecomment-2569952406 (apply that patch)
* undo `        results = await asyncio.gather(*tasks, return_exceptions=True)`

it will resolve both issues.

akoeplinger committed:
Add signature checking to install-debs.py

akoeplinger committed:
Fix unassigned variable error

akoeplinger committed:
WIP

akoeplinger committed:
Merge remote-tracking branch 'upstream/main' into akoeplinger-patch-1

# Conflicts:
#	eng/common/cross/build-rootfs.sh

akoeplinger committed:
Improve --keyring handling, remove unused armel/armv6 cases

akoeplinger committed:
Add armel handling back for tizen-build-rootfs.sh

akoeplinger committed:
Switch tar extraction filter back to filter='tar' instead of 'fully_trusted'

I can't reproduce the problem I hit with https://github.com/dotnet/arcade/pull/15372#issuecomment-2569813723 anymore and this is safer.

akoeplinger committed:
Harden install-debs.py error handling and shell invocation

akoeplinger committed:
Make --mirror required for install-debs.py

We always pass it in build-rootfs.sh and it removes some duplication

akoeplinger committed:
Consistently use https URLs for added hardening
